Step3 PR

KaKaFE opened the pull request:
Step3 PR 입니다.
commit을 거의 못했습니다... 다음 스텝 에선 의식하고 하겠습니다.
transition 을 이용해서 plans 레이어를 구현하였고
JS 는 class를 이용해서 만들 었습니다....코드가 깔끔 하지 못한 것 같습니다 ㅠㅠ 
많은 피드백 부탁드립니다...

demo : https://kakafe.github.io/amazoncopy/

Commit:
menulist의 오타 수정

Commit:
main.css 리팩토링

중복된 값들을 묶어 가독성을 높임, nav-top부분에 wrap 을 만들어, page크기가 달라졌을때,
가운데 정렬되도록 수정

Commit:
step3 최종기능 구현

스크롤 했을때, plan page 가 펼처지고, 그안에서의 인터렉션 구현 CSS파일을 section별로 구분,
nav.js파일생성 및 기능구현

crongro commented:
데모 잘 봤어요~ 

커밋로그는 습관들이기 어렵긴해요. 그래도 계속 의식해보시고요.

코드가 깔끔하지 못하는 것에 너무 의식하지 마시고,  나중에 어떻게 개선하면 되겠군. 이런생각 정도 하셔도 좋아요.
좋은코드는 정답도 끝도 없어서.. 너무 고민에 빠지면 힘들어요.

crongro commented on CSS/main.css:
```diff
@@ -0,0 +1,21 @@
+/* 기본 CSS */
+
+* {
```
css역시 공통 부분을 이렇게 분리하는 건 좋습니다.

crongro commented on CSS/nav-planpage.css:
```diff
@@ -0,0 +1,140 @@
+/* more plan page Style */
+
+.planpage {
```
이름이 어느정도 함축적이여도 좋아요.
너무 길면 가독성이 오히려 떨어져서... 
예를들어 planpage보다는 plan 이라고 해도 뭔지 알거 같네요.

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,47 @@
+class Nav_Elements {
```
네이밍에, 
카멜표기법과 (_)를 혼용할 필요는 없을 듯.

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,47 @@
+  }
+  addMorePlanDownSizeEvent() {
+    this.moreCntBtn.addEventListener("click", this.removeClass.bind(this));
+    this.moreCntArrow.addEventListener("click", this.removeClass.bind(this));
```
arrow function을 쓰면 지금 bind가 필요없죠? 왜 그런지도 알고 계시면 좋고요.

KaKaFE replied:
처음에는 arrow function을 사용해서 구현했는데, addEventListener 내부 코드가 길어져서 따로 빼서 bind를 했는데, arrow function을 활용하는 다른 방법이 있을까요?? 궁금합니다~!

crongro replied:
bind를 해도 돼요. arrow함수를 쓸수도 있다는 것이고요.
arrow function을 쓰면 왜 bind를 안해도 되는지를 이해하셔야하고요.

```javascrit
this.moreCntBtn.addEventListener("click", (e)   => this.removeClass(e));
```

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,47 @@
+class Nav_Elements {
+  constructor() {
+    this.planPage = document.querySelector(".planpage");
```
constructor에  너무 많은 this property가 없으면 좋겠어요.
변수(상태)는 유효범위를 좁히는 게 중요한데요.  
class property보다는 각 메서드의 지역변수로 옮길 수 있으면 옮기도록 하세요.

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,47 @@
+    el.classList.remove(className);
+  }
+  removeClass() {
+    this.removeClassList(this.planPage, "planpage-more-click");
```
planpage-more-click이런 정보를 객체로 만들어서 쓰는 건 어때요?

const CLASS_MAP = {  "more" : "planpage-more-click",....}

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,47 @@
+}
+
+const NavElements = new Nav_Elements();
+NavElements.addScrollEvent();
```
이 세줄을 밖에서 해도 되는데요.
그것보단,  Nav_Elements  클래스 안에서 initialize() 메서드 같은거 만들어서 모아두고, 

이부분에서는 NavElements.initialize()  메서드를 호출하는 건 어때요?

crongro reviewed (changes requested):
수고하셨어요. 👍 

전체 구조와 이벤트, DOM이해등이 괜찮아 보여요.

리뷰남겨드린 거 한번 수정하고 해보시죠.

Commit:
planpage class명 수정 및 nav.js의 변수명을 수정

가독성 향상을 위해 planpage를 plan으로 변경, Nav_Elements 를 NavElements로 수정함

Commit:
navElements 의 프로퍼티를 각메서드의 지역변수로 옮긴다

2개이상의 메서드에서 사용되는 프로퍼티를 제외하고 나머지를 addMoreClickEvent 메서드와
scrollHandler,addMorePlanDownSizeEvent메서드에 지역변수로 넣는다

Commit:
classMap 객체를 생성

classMap 객체를 생성하여, 변수명을 classMap 객체를통해 맵핑하는 방식으로 변경한다

Commit:
initInteraction 메서드를 생성

기존 3개의 메서드를 실행시켜 인터렉션을 작동하는데, initInteraction메서드로 묶어 한번에 실행

KaKaFE commented:
리뷰 주신 사항 수정하였습니다!
-질문
리뷰 주신 것 중에, class map 객체를 통해서, 맵핑 하는 방식을 구현할때,
navElements 클래스 안에 프로 퍼티로 넣자니 접근 할때  navElements .classMap.more  이렇게 
길어질거 같아서, 전역변수로 만들었는데요,  어떤게 더나은 방법인지 궁금합니다

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,58 @@
+  }
+}
+
+const classMap = {
```
classMap을 주입시키는 것도 가능할 듯. 

```javascript
const navElement = new NavElements(classMap);

...

class NavElements {
  constructor(cmap) {  //받을땐 이름을 좀 줄여써도 되고요.
    this.cmap = cmap;
    this.el.addClass(this.cmap.middleIntrc);
  }
```

그리고 ES Modules를 활용하면 classMap을 별도 파일로 분리하고, import받아서 쓸수도 있고요.

crongro commented on main.html:
```diff
@@ -5,112 +5,150 @@
-    <link rel="stylesheet" type="text/css" href="/main.css" />
+    <link rel="stylesheet" type="text/css" href="./CSS/main.css" />
+    <link rel="stylesheet" type="text/css" href="./CSS/nav.css" />
+    <link rel="stylesheet" type="text/css" href="./CSS/nav-plan.css">
```
실무에서는 css파일도 여러개 배포하지 않고, 개발할때는 분리하지만 배포할때는 하나로 합치죠.
HTTP Request수를 줄이기 위함이에요. (로딩 성능개선의 기본)

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,58 @@
+    this.plan = document.querySelector(".plan");
+    this.morePage = document.querySelector(".plan-more-contents");
+    this.planCnt = document.querySelector(".plan-contents");
+    this.navPosition = 112;
```
이런값은 계산에 의해서 값을 얻거나...아니면 NavElements클래스를 초기활할때 인자로 받거나 하세요.
이렇게 하드코딩하는 건 별로에요~ 

new NavElemtns({ navPosition : 112, primBtnPosition: 465}) 이런식도 좋죠.

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,58 @@
+    this.navPosition = 112;
+    this.primBtnPosition = 465;
+  }
+  removeClassList(el, className) {
```
이 함수가 필요없다고 생각하고요. 한줄짜리 함수라서...
더구나 클래스를 제거하는 것인데,  이름만 봐서는 classlist를 제거한다고 되어 있네요.

crongro commented on JS/nav.js:
```diff
@@ -0,0 +1,58 @@
+    this.morePage.classList.add(classMap.click);
+    this.planCnt.classList.add(classMap.donwSize);
+  }
+  initInteraction() {
```
Interaction인데 스펠링이 틀린 듯~

crongro commented on CSS/nav-plan.css:
```diff
@@ -0,0 +1,140 @@
+}
+
+.more-contents-img {
+	width: 998px;
```
이제부터 CSS구현하실때
em이나 rem을 공부하시고요.

px이 필요한 곳(font, box-model)에 rem을 사용해보세요~

crongro reviewed (approved):
네 수정확인했어요.
클래스나 메서드의 크기는 좋습니다. 이렇게 개발하시고요.
클래스를 사용해도 함수의 인자를 좀더 많이 활용하면 좋겠어요.그래야 testable  code를 만들 수 있습니다.

리뷰참고해서 다음과제 개발해보세요~
